### PR TITLE
zpi-77 - flattening of the role graph

### PR DESCRIPTION
**Pull request by Nightbaka**
Forget about graphs, its time for matrices

### CODE REVIEW

**Comment by Nightbaka** [comment]
I've changed my mind, I will resubmit this PR changes when I have the entire endpoint ready, to not make you go through this multiple times

**Review by samuelus — CHANGES_REQUESTED** [review]
I'll do a review later, but I want to have dibs so it doesn't go without my approval. Btw please don't merge it with this name XD

File: backend/auth/role_map_repository_test.go
Comment (Ninja00Shadow) [inline backend/auth/role_map_repository_test.go:694]: What is this? To make tests easier?

File: backend/auth/role_map_repository.go
Comment (Ninja00Shadow) [inline backend/auth/role_map_repository.go:19]: This is a nightmare fuel.

**Review by Ninja00Shadow — COMMENTED** [review]
This is too much for me.

File: backend/auth/role_map_repository_test.go
Comment (Nightbaka) [reply backend/auth/role_map_repository_test.go:694]: This is generating a random role map, with given parameters. Its used for fuzzy tests - tests with random inputs. I decided this approach would be better than hardcoded test cases to avoid missing out any edge cases

File: backend/auth/role_map_repository.go
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:19]: you can think of it as a map of role -> permission matrix

Comment (Nightbaka) [reply backend/auth/role_map_repository.go:19]: I had no nightmares about that while coding so can't confirm this is a valid comment

File: backend/auth/role_map_repository.go
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:239]: Yes but I would rather keep these 2 separate. I could refactor this into function accepting func operating on operations, but I would rather have them like this

Comment (Ninja00Shadow) [reply backend/auth/role_map_repository.go:239]: Can you at least separate this to another function? This should be easy:

```
_, hasNamespace := matrix[permit.Namespace]
	_, hasResource := matrix["*"][permit.Resource]

	if !hasNamespace {
		expandNamespaces(permit.Namespace, matrix)
	}
	if !hasResource {
		expandResources(permit.Resource, matrix)
	}
```

File: backend/auth/role_map_repository_test.go
Comment (Ninja00Shadow) [reply backend/auth/role_map_repository_test.go:694]: Understandable, have a great day

File: backend/auth/role_map_repository.go
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:239]: I would prefer not to do that, I think it would actually worsen, the readability. I will add 6 line shorter version instead. Will that suffice?

**Review by Ninja00Shadow — APPROVED** [review]

**Comment by Ninja00Shadow** [comment]
image

File: backend/auth/role_map_repository.go
Comment (samuelus) [reply backend/auth/role_map_repository.go:19]: It looks like a nightmare, but under the hood it's brilliant. I really admire you for finding this optimization

File: backend/auth/role_map_repository_test.go
Comment (samuelus) [inline backend/auth/role_map_repository_test.go:1]: They have high coverage so i will assume they are good xd

Comment (samuelus) [reply backend/auth/role_map_repository_test.go:1]: But you can add this case to TestHasCycle, to get full coverage of this function. In the future for coverage raport you can use this commands:
```
go test -coverprofile cover.out
go tool cover -html=cover.out
```

```
,
		{
			name: "Non-existing subrole",
			roleMap: map[string]*models.Role{
				"admin": {Subroles: []string{"nonexistent"}},
			},
			expected: false,
		},
```

**Review by samuelus — CHANGES_REQUESTED** [review]
I know I gave a lot of comments, but I'm really impressed how good this code is, looking at its complexity

File: backend/auth/role_map_repository.go
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:35]: Sure, there are more detailed errors logged in the GetRoleMapConfig anyways.

File: backend/auth/role_map_repository.go
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:134]: There is no point besides readability and writability. This is too clutered with syntax for my taste:
`if len(m1) * len(m1["*"]) < len(m2) * len(m2["*"])  {
    m1, m2 = m2, m1
}`

File: backend/auth/role_map_repository.go
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:233]: Ehhh it could be I just still can't believe the same exact logic for both of these works

File: backend/auth/role_map_repository.go
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:278]: I need it for comparison tests atleast. Was thinking of using it as fallback if initializing the matrix would take a longer time then expected to, but honestly propably won't do that. Something about moving it to the test file doesn't sit right in me, but if you insist it could be done

File: backend/auth/role_map_repository.go
Comment (Ninja00Shadow) [reply backend/auth/role_map_repository.go]: > Yes but I would rather keep these 2 separate. I could refactor this into function accepting func operating on operations, but I would rather have them like this

Traitor.

Comment (Nightbaka) [reply backend/auth/role_map_repository.go]: Samuel gave me a ready refactor,  you just yyyyyy... can it be simplified?

Comment (Nightbaka) [reply backend/auth/role_map_repository.go]: Yes I was lazy

File: backend/auth/role_map_repository.go
Comment (samuelus) [reply backend/auth/role_map_repository.go:246]: If it is a function used only in tests, then it should be in the test file, because now it looks like it is not used anywhere (for tests I just thought it is used in its tests, not in their creation), plus it spoils the test coverage, as if someone would ever look at it more

**Review by samuelus — APPROVED** [review]

**Pull request merged**
